vnext/Mso/future/futureWinRT.h

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+#ifndef MSO_FUTURE_FUTUREWINRT_H
+#define MSO_FUTURE_FUTUREWINRT_H
+
+#include <errorCode/hresultErrorProvider.h>
+#include <future/future.h>
+#include <winrt/Windows.Foundation.h>
+
+namespace Mso {
+
+struct AsyncActionFutureAdapter : winrt::implements<
+                                      AsyncActionFutureAdapter,
+                                      winrt::Windows::Foundation::IAsyncAction,
+                                      winrt::Windows::Foundation::IAsyncInfo> {
+  using AsyncStatus = winrt::Windows::Foundation::AsyncStatus;
+  using AsyncCompletedHandler = winrt::Windows::Foundation::AsyncActionCompletedHandler;
+
+  AsyncActionFutureAdapter(Mso::Future<void> &&future) noexcept : m_future{std::move(future)} {
+    m_future.Then<Mso::Executors::Inline>([weakThis = get_weak()](Mso::Maybe<void> && result) noexcept {
+      if (auto strongThis = weakThis.get()) {
+        AsyncCompletedHandler handler;
+        {
+          winrt::slim_lock_guard const guard(strongThis->m_lock);
+          if (strongThis->m_status == AsyncStatus::Started) {
+            if (result.IsValue()) {
+              strongThis->m_status = AsyncStatus::Completed;
+              if (strongThis->m_completedAssigned) {
+                handler = std::move(strongThis->m_completed);
+              }
+            } else {
+              strongThis->m_status = AsyncStatus::Error;
+              strongThis->m_error = result.GetError();
+            }
+          }
+        }
+
+        if (handler) {
+          invoke(handler, *strongThis, AsyncStatus::Completed);
+        }
+      }
+    });
+  }
+
+  void Completed(AsyncCompletedHandler const &handler) {
+    AsyncStatus status;
+
+    {
+      winrt::slim_lock_guard const guard(m_lock);
+
+      if (m_completedAssigned) {
+        throw winrt::hresult_illegal_delegate_assignment();
+      }
+
+      m_completedAssigned = true;
+
+      if (m_status == AsyncStatus::Started) {
+        m_completed = winrt::impl::make_agile_delegate(handler);
+        return;
+      }
+
+      status = m_status;
+    }
+
+    if (handler) {
+      invoke(handler, *this, status);
+    }
+  }
+
+  auto Completed() noexcept {
+    winrt::slim_lock_guard const guard(m_lock);
+    return m_completed;
+  }
+
+  uint32_t Id() const noexcept {
+    return 1;
+  }
+
+  AsyncStatus Status() noexcept {
+    winrt::slim_lock_guard const guard(m_lock);
+    return m_status;
+  }
+
+  winrt::hresult ErrorCode() noexcept {
+    try {
+      winrt::slim_lock_guard const guard(m_lock);
+      RethrowIfFailed();
+      return 0;
+    } catch (...) {
+      return winrt::to_hresult();
```

Comment:
Does this assume the exception is an hresult_erorr? Wondering if we should catch only that instead of catch all.